src/main/java/jenkins/scm/api/SCMNavigator.java

```diff
@@ -64,12 +66,92 @@
             = new AlternativeUiTextProvider.Message<SCMNavigator>();
 
     /**
+     * Cache of the ID of this {@link SCMNavigator}.
+     * @since FIXME
+     */
+    private transient String id;
+
+    /**
      * Constructor.
      */
     protected SCMNavigator() {
     }
 
     /**
+     * Returns the ID of the thing being navigated.
+     * <p>
+     * The ID will typically be a composite of things like the server and the project/organization that the navigator
+     * is scoped to.
+     * <p>
+     * For example, a GitHub navigator that is navigating repositories in a GitHub organization could construct
+     * its ID as being the URL of the GitHub Server (to allow for GitHub Enterprise servers) and the name of the
+     * organization.
+     * <p>
+     * The key criteria is that if two navigators have the same ID <strong>and</strong> they are both in the same
+     * {@link SCMNavigatorOwner} then the results from
+     * {@link #fetchActions(SCMNavigatorOwner, SCMNavigatorEvent, TaskListener)} should be not just equivalent but
+     * {@link List#equals(Object)}.
+     *
+     * @return the ID of the thing being navigated by this navigator.
+     * @since FIXME
+     * @see #id()
+     */
+    @NonNull
+    public final String getId() {
+        if (id == null) {
+            if (MethodUtils.isAbstract(getClass(), "id")) {
+                // we need to ensure that upgrading the plugin does not change the digest
+                // processing XML with a regex is usually a bad thing. In this case we have a sufficiently
+                // strict set of requirements to match (shortname)@(version) and the constraints on the
+                // shortname and version are such that we can have a high degree of certainty of a match
+                // and even if we glom some legitimate text, the chances of a duplicate that is otherwise
+                // the same but has a critical difference in a regular value that happens to contain
+                // plugin="xys@abc" is sufficiently low.
+                id = getClass().getName() + "::" + Util.getDigestOf(Items.XSTREAM.toXML(this)
+                        .replaceAll(" plugin=(('[^']+@[^']+')|(\"[^\"]+@[^\"]+\"))", ""));
+            } else {
+                id = getClass().getName() + "::" + id();
+            }
+        }
+        return id;
+    }
+
+    /**
+     * If implementations are using {@link DataBoundSetter} on fields that affect the {@link #id()} calculation then
+     * those fields must call {@link #resetId()} if they may have invalidated the cached {@link #getId()}.
```

Comment:
IMO better to leave the choice of whether a cache is needed up to the implementation, and just have

```java
public String getId() {
    return getClass().getName() + "::" + Util.getDigestOf(Items.XSTREAM.toXML(this)
        .replaceAll(" plugin=(('[^']+@[^']+')|(\"[^\"]+@[^\"]+\"))", ""));
}
```

After all, the current system will not correctly `resetId` when there are configuration changes in implementations failing to override `id`.